### ISSUE TITLE
Failed to run MNIST example on Android

### ISSUE THREAD

**Issue by guojia-git** [tensorflow/tensorflow#5338]
### What related GitHub issues or StackOverflow threads have you found by searching the web for your problem?
#4623
#4451

### Environment info
Operating System: Ubuntu 14.04
If installed from source, provide 

1. The commit hash (`git rev-parse HEAD`): 3d35376a66cde4f3e614c746d3c8708d15caa1b5
2. The output of `bazel version`: Build label: 0.3.2

### If possible, provide a minimal reproducible example (We usually don't have time to read hundreds of lines of your code)

- Train using [MNIST tutorial](https://www.tensorflow.org/versions/r0.11/tutorials/mnist/pros/index.html)
- `tf.train.write_graph(..., mnist_graph.txt, as_text=True)`
- freeze graph `bazel-bin/.../freeze_graph --input_graph=.../mnist_graph.txt --input_checkpoint=.../mnist_model.ckpt --output_graph=.../mnist_graph_frozen.pb --output_node_names="output_node"`
- strip graph `bazel-bin/..bazel-bin/.../strip_unused --input_graph=.../mnist_graph_frozen.pb --output_graph=.../mnist_graph_frozen_stripped.pb --input_node_names="input_node" --output_node_names="output_node" --input_binary=true`
- Java code: basically I took part of TensorFlowImageListener and part of TensorFlowImageClassifier
  - Remove the image preprocessing part, instead pass in the raw mnist datapoint of shape [784] as floatValues
  - `inferenceInterface.fillNodeFloat(”input_node", 1, 28, 28, 1, floatValues); //When training I was reshaping the datapoint to 28x28. `
  - `inferenceInterface.runInference(new String[] {"output_node"})`
  - `inferenceInterface.readNodeFloat("output_node", outputs")`

### Logs or other output that would be helpful
(If logs are large, please upload as attachment or provide link).
10-23 19:16:51.918: I/native(10721): tensorflow_inference_jni.cc:126 Creating session.
10-23 19:16:51.967: I/native(10721): tensorflow_inference_jni.cc:137 Tensorflow graph loaded from: file:///android_asset/mnist_graph_frozen_stripped.pb
10-23 19:16:51.967: I/native(10721): tensorflow_inference_jni.cc:140 Initialization done in 137.556ms
10-23 19:16:51.967: I/MainActivity(10721): Tensorflow model initialized
10-23 19:16:51.988: I/native(10721): tensorflow_inference_jni.cc:87 Found session variables for 6f51f9109840554e
10-23 19:16:51.988: I/native(10721): tensorflow_inference_jni.cc:87 Found session variables for 6f51f9109840554e
10-23 19:16:52.194: I/native(10721): tensorflow_inference_jni.cc:197 End computing. Ran in 205ms (205ms avg over 1 runs)
**10-23 19:16:52.194: E/native(10721): tensorflow_inference_jni.cc:202 Error during inference: Invalid argument: You must feed a value for placeholder tensor 'Placeholder' with dtype float
10-23 19:16:52.194: E/native(10721): 	 \[\[Node: Placeholder = Placeholder \[dtype=DT_FLOAT, shape=[], _device="/job:localhost/replica:0/task:0/cpu:0"]()]]**
10-23 19:16:52.194: I/native(10721): tensorflow_inference_jni.cc:87 Found session variables for 6f51f9109840554e
10-23 19:16:52.194: I/native(10721): tensorflow_inference_jni.cc:87 Found session variables for 6f51f9109840554e
**10-23 19:16:52.195: E/native(10721): tensorflow_inference_jni.cc:159 Output [output_node] not found, aborting!**

**Thanks for reading this. Please let me know if there's any other information I need to provide.**

**Comment by aselle** [comment]
I would try loading the graph in python and see if you can feed it there to see if the problem is with your baked trained model or with your Java usage. @petewarden might have more precise suggestions. Could there be an errant Placeholder that is in the mnist_model you created that you could just feed with a dummy value just to get the ball rolling?

**Comment by guojia-git** [comment]
@aselle Thanks for the suggestions! I haven't fully sort it out yet, but yes you are right. The 'Placeholder' was referring to the drop out layer that's not given a name. Will update and close the issue once I fix it.

**Comment by aselle** [comment]
Thanks for the update, good luck.

**Comment by guojia-git** [comment]
@aselle I finally got it running.

First I tried to work around with that dropout layer in the model, by feeding the `Placeholder` (the keep probability) with a value `{1f}`, but I ended up getting `In[0] is not a matrix` at the `MatMul `that uses the dropout as input. I gave up after trying a few things to solve it.

The final working model was a clean one with manually assigned parameters.

I guess understanding the error messages has been the biggest obstacle for me throughout the process. It is especially hard for people who are not working with the C++ API (like me). Something like `You must feed a value for placeholder tensor XXX` happens when some placeholder exists in the graph but was not fed with a value, while `FeedInputs: unable to find feed output` happens when some placeholder does not exist in the graph but you are feeding value to it. Thanks again for helping out. Closing the issue.
